homeassistant/components/caldav/config_flow.py

```diff
@@ -0,0 +1,170 @@
+    CONF_CUSTOM_CALENDARS,
+    CONF_DAYS,
+    DEFAULT_VERIFY_SSL,
+    DOMAIN,
+)
+
+DATA_SCHEMA = vol.Schema(
+    {
+        # pylint: disable=no-value-for-parameter
+        vol.Required(CONF_USERNAME): selector.TextSelector(),
+        vol.Required(CONF_PASSWORD): selector.TextSelector(
+            selector.TextSelectorConfig(type=selector.TextSelectorType.PASSWORD)
+        ),
+        vol.Required(CONF_URL): selector.TextSelector(
+            selector.TextSelectorConfig(type=selector.TextSelectorType.URL)
+        ),
+        vol.Optional(
+            CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL
+        ): selector.BooleanSelector(),
+        vol.Optional(CONF_DAYS, default=1): cv.positive_int,
+        vol.Optional(CONF_CALENDARS, default=[]): selector.SelectSelector(
+            selector.SelectSelectorConfig(options=[], custom_value=True, multiple=True),
+        ),
+        vol.Required(CONF_ADD_CUSTO_CALENDAR, default=False): bool,
+    }
+)
+
+CUSTOM_CALENDARS_SCHEMA = vol.Schema(
+    {
+        vol.Optional(
+            CONF_CUSTOM_CALENDARS,
+            description={
+                "suggested_value": [
+                    {
+                        "name": "HomeOffice",
+                        "calendar": "Agenda",
+                        "search": "HomeOffice",
+                    },
+                    "# Please clean this example",
+                ]
+            },
+        ): selector.ObjectSelector(),
+    }
+)
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class CaldavFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a caldav config flow."""
+
+    VERSION = 1
+
+    def __init__(self):
+        """Initialize."""
+        self.user_input = None
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get option flow."""
+        return CaldavOptionsFlowHandler(config_entry)
+
+    async def async_step_user(self, user_input=None) -> FlowResult:
+        """Handle a flow initialized by the user."""
+        errors = {}
+        if user_input is not None:
+            await self.async_set_unique_id(
+                f"{user_input[CONF_USERNAME]}:{user_input[CONF_URL]}"
+            )
+            self._abort_if_unique_id_configured()
+
+            if user_input.get(CONF_ADD_CUSTO_CALENDAR):
+                return await self.async_step_custom_calendars(user_input)
+
+            try:
+                await async_caldav_connect(self.hass, user_input)
+                return self.async_create_entry(
+                    title=DOMAIN,
+                    data=user_input,
+                    options={
+                        CONF_CALENDARS: user_input.pop(CONF_CALENDARS, []),
+                        CONF_CUSTOM_CALENDARS: user_input.pop(
+                            CONF_CUSTOM_CALENDARS, []
+                        ),
+                    },
+                )
+
+            except Exception as exception:  # pylint:disable=broad-except
```

Comment:
We should catch the connection errors only here.  

Its ok to use a broad exception catch as a fallback and set the error to unknown here but only after you have caught error we know are connection problems